API: Account changes to event stream

kfangw opened the pull request:
### Requirements

* Implement SSE(Sever Sent Event)

### Github Issue
#3 

### Solution
I changed two point
- server-sent-event api with go-observerble
- Adopt mux

After this change, i will add bellows
- client
- APIs for each issues

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bosnet/sebak?pullRequest=84) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bosnet/sebak?pullRequest=84) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bosnet/sebak?pullRequest=84) it.</sub>

kfangw commented:
@spikeekips @zzim2x Please review

outsideris commented on lib/broker_test.go:
```diff
@@ -0,0 +1,46 @@
+	ret2 := buffer2.String()
+
+	fmt.Println(ret1)
+	fmt.Println(ret2)
```
These are unnecessary.

kfangw replied:
yeap. I will remove it.

outsideris commented on lib/broker.go:
```diff
@@ -0,0 +1,142 @@
+const (
+	EventDummy  EventType = "dummy"
+	EventDummy2           = "dummy2"
+)
```
Why is this needed? `dummy` looks like test purpose.

kfangw replied:
Event types are not yet defined. Hence, I added those two.
Maybe those are changed to practical names

anarcher commented on lib/broker.go:
```diff
@@ -0,0 +1,142 @@
+)
+
+type EventBroker struct {
+	brokers map[EventType]*Broker
```
IMHO, `EventBroker` seems like  `EventBrokers`.

anarcher commented on lib/broker.go:
```diff
@@ -0,0 +1,142 @@
+	e := &EventBroker{
+		brokers: map[EventType]*Broker{},
+	}
+	e.brokers[EventDummy] = NewBroker()
```
IMHO, This statement `e.brokers[EventDummy] = ..` is not necessary.

kfangw replied:
AddBroker much better 👍

kfangw commented:
I am working on another testcase.. the testcase may fail.

spikeekips commented:
I understand why you invented `EventBroker`. I thought you already deeply considered the other implementation of internal event systems. We can make it easier like https://github.com/GianlucaGuarini/go-observable or https://github.com/ReactiveX/RxGo .

cmcm2222 commented:
Wow. 
Reactive programming can be in golang too?? 
Awesome.

kfangw commented:
Good to know those projects.
I will adopt one one them or make a similar one. Thanks spike!!

kfangw commented:
@spikeekips @zzim2x @outsideris @anarcher 
please review again.
At this moment, please ignore broken ci build. The broken caused by CI itself

I changed two point
1. server-sent-event api with go-observerble
2. Adopt mux

After this change, i will add bellows
1. client
2. APIs for each issues

spikeekips commented:
@kfangw Please ask us after finishing work :) Too often changes, I am confused when I start to review.

kfangw commented:
Writing is done but there is a CI problem...T^T...
Sorry for the annoying...
I will ask again when the problem is solved

spikeekips commented:
@kfangw Thanks :)

kfangw commented:
Lets move to https://github.com/bosnet/sebak/pull/114

kfangw commented:
CI build failed occurred occasionally.
However, the original HANG problem is not seen anymore.
I am sorry for the annoying until now.

Please review this PR

outsideris commented:
DId you find what caused the hang? just curious.

kfangw commented:
circleci ran the build with old source.
When I submit another PR with the same branch, It turns to normal...
I don't know what happened in the CI exactly.

kfangw commented:
@outsideris 
refer this https://circleci.com/gh/bosnet/sebak/190
At that time, I removed the testcase "TestStreaming"
but it ran test with old source.

I guess that the CI couldn't fetch latest source code. 

I have trying to reproduce the problem, but I can't...
Unfortunately, I squashed the commit and force pushed... so i lost the commit T^T

anarcher commented on lib/api.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func GetAccountHandler(ctx context.Context) HandlerFunc {
+	storage, ok := ctx.Value("storage").(*sebakstorage.LevelDBBackend)
```
IMHO,  Using `LevelDBBackend` with `context.Context.Value` is not suitable in normal cases.

kfangw replied:
Issue what you picked is spread all around of handlers for sebak.
I think we need to talk about this in another issue.

anarcher commented on lib/api_test.go:
```diff
@@ -0,0 +1,95 @@
+	ctx := context.WithValue(context.Background(), "storage", storage)
+	router := mux.NewRouter()
+	router.HandleFunc("/account/{address}", GetAccountHandler(ctx)).Methods("GET")
+	server := &http.Server{Addr: ":5000", Handler: router}
```
IHMO, Using `net/http/httptest` can be better for using testing :-D.

kfangw replied:
I tried the httptest.
May I change when I add another APIs?

anarcher replied:
Using httptest is optional :-)!

anarcher commented on lib/block_account.go:
```diff
@@ -42,6 +43,8 @@ func (b *BlockAccount) String() string {
 	return string(sebakcommon.MustJSONMarshal(b))
 }
 
+var BlockAccountObserver = observable.New()
```
IHMO, In general, **package level vars** as using  **Package-global objects** can have surprising side effects.  Anyway, This var can be in another place. :)

kfangw replied:
I don't like this usage also. I am totally agreed what you worry about.
However, the object should be accessed from everywhere.
So, I will just move it to another separated package.
Is it enough? or is there any alternative solution? @anarcher

anarcher replied:
@kfangw  I understand why it has to exist in pkg-global var. It makes sense :-)

kfangw commented on lib/api.go:
```diff
@@ -0,0 +1,119 @@
+
+func AddAPIHandlers(ctx context.Context, t *sebaknetwork.HTTP2Network) {
+
+	t.AddAPIHandler(ctx, "/account/{address}", GetAccountHandler).Methods("GET")
```
I think that we may need a bunch of features provided by mux package.
hence, we need `*mux.Route` reference here.

@spikeekips

kfangw commented:
@spikeekips @anarcher May I merge this PR?
remaining issues pointed in this PR are ... 

1. unittest with httptest
2. usage of context

those will be solved in another PR.

anarcher commented:
IMHO, unittest with httptest is optional. You can ignore it!! :-D

anarcher commented on lib/api.go:
```diff
@@ -0,0 +1,63 @@
+	"boscoin.io/sebak/lib/network"
+	"github.com/GianlucaGuarini/go-observable"
+)
+
```
I know my approach is ugly. `T_T`
```
⚠ func (nr *NodeRunner) Ready() {
      nr.network.SetContext(nr.ctx)
      nr.network.AddHandler(nr.ctx, AddAPIHandlers(nr.storage))
      nr.network.Ready()
  }
```

```
⚠ func AddAPIHandlers(s *sebakstorage.LevelDBBackend) func(ctx context.Context, t *sebaknetwork.HTTP2Network) {
      fn := func(ctx context.Context, t *sebaknetwork.HTTP2Network) {
          //t.AddAPIHandler(ctx, "/account/{address}", GetAccountHandler).Methods("GET")
          t.AddAPIHandler("/account/{address}", GetAccountHandler(ctx, t, s)).Methods("GET")
      }
      return fn
  }
```

```
func GetAccountHandler(ctx context.Context, t *sebaknetwork.HTTP2Network, s *sebakstorage.LevelDBBackend) http.Handle
  rFunc {
      /*
          storage, ok := ctx.Value("storage").(*sebakstorage.LevelDBBackend)
          if !ok {
              panic(errors.New("storage is missing in context"))
          }
      */

      storage := s
```

```
⚠ func (t *HTTP2Network) AddAPIHandler(pattern string, handlerFunc http.HandlerFunc) (router *mux.Route) {
      //pattern := args[0].(string)
      //handler := args[1].(func(context.Context, *HTTP2Network) HandlerFunc)
      apiRouter := t.routers["api"]
      return apiRouter.HandleFunc(pattern, handlerFunc)
  }
```

kfangw replied:
It is not ugly. elegance. 👍 
I have been suffered from circular import, hence my dirty code been written.
Your idea makes me free from circular import and refer to the storage
Thanks @anarcher

anarcher replied:
@kfangw  I am honored that I have helped you. `:-]`

spikeekips commented on lib/network/http2_network.go:
```diff
@@ -61,15 +69,22 @@ func NewHTTP2Network(config HTTP2NetworkConfig) (h2n *HTTP2Network) {
+	h2n.handlers = map[string]func(http.ResponseWriter, *http.Request){}
+	h2n.routers = map[string]*mux.Router{
+		"node": baseRouter.PathPrefix("/node").Subrouter(),
+		"api":  baseRouter.PathPrefix("/api").Subrouter(),
```
We can use `UrlPrefixXX` in this place even in map key.

kfangw replied:
Yeap. I changed it to named const.

spikeekips commented on lib/network/http2_network.go:
```diff
@@ -10,25 +10,33 @@ import (
+	UrlNodePrefix = "/node"
+	UrlApiPrefix  = "/api"
+)
+
```
In the below code for `HTTP2Network.routers`, you can define url path prefix and router name too.
```go
const (
	RouterNameNode = "node"
	RouterNameAPI  = "node"
)

var (
	UrlPathPrefixNode = fmt.Sprintf("/%s", RouterNameNode)
	UrlPathPrefixAPI  = fmt.Sprintf("/%s", RouterNameAPI)
)
```

spikeekips commented:
@kfangw I think this PR is the 1st implementation of API using **server side stream**, so how about change the PR title for the original issue, #3 .
